File: src/decoders/resource.ts
import type { SessionHandle, Resource } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { decodeSubject } from "./subject";
import { decodeResourceContent } from "./resource-content";

export const decodeResource = (resource: any, session: SessionHandle): Resource => {
  return {
    id: resource.N,
    startDate: decodePronoteDate(resource.Date.V),
    endDate: decodePronoteDate(resource.DateFin.V),
    subject: decodeSubject(resource.Matiere.V),

    haveAssignment: typeof resource.dateTAF !== "undefined",
    assignmentDeadline: resource.dateTAF?.V && decodePronoteDate(resource.dateTAF.V),

    teachers: resource.listeProfesseurs.V.map((teacher: any) => teacher.L),
    groups: resource.listeGroupes.V.map((group: any) => group.L),

    backgroundColor: resource.CouleurFond,
    contents: resource.listeContenus.V.map((content: any) => decodeResourceContent(content, session))
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/resource.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.Resource
import ink.literate.pawnote.models.SessionHandle
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

fun decodeResource (resource: JsonObject, session: SessionHandle) = Resource(
    id = resource["N"]!!.jsonPrimitive.content,
    startDate = decodePronoteDate(resource["Date"]!!.jsonObject["V"]!!.jsonPrimitive.content),
    endDate = decodePronoteDate(resource["DateFin"]!!.jsonObject["V"]!!.jsonPrimitive.content),
    subject = decodeSubject(resource["Matiere"]!!.jsonObject["V"]!!.jsonObject),

    haveAssignment = resource.contains("dateTAF"),
    assignmentDeadline = if (resource.contains("dateTAF")) decodePronoteDate(resource["dateTAF"]!!.jsonObject["V"]!!.jsonPrimitive.content) else null,

    teachers = resource["listeProfesseurs"]!!.jsonObject["V"]!!.jsonArray.map { it.jsonObject["L"]!!.jsonPrimitive.content },
    groups = resource["listeGroupes"]!!.jsonObject["V"]!!.jsonArray.map { it.jsonObject["L"]!!.jsonPrimitive.content },

    backgroundColor = resource["CouleurFond"]!!.jsonPrimitive.content,
    contents = resource["listeContenus"]!!.jsonObject["V"]!!.jsonArray.map { decodeResourceContent(it.jsonObject, session) }
)